sw/device/lib/testing/usb_logging.c

```diff
@@ -0,0 +1,549 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/testing/usb_logging.h"
+
+#include <ctype.h>
+
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/lib/testing/pinmux_testutils.h"
+#include "sw/device/lib/testing/usb_testutils.h"
+#include "sw/device/lib/testing/usb_testutils_controlep.h"
+
+#define MODULE_ID MAKE_MODULE_ID('u', 's', 'l')
+
+// Size of each logging buffer; data is collected in a buffer until the buffer
+// is full (or a timeout occurs) and is then transmitted as a single transfer
+// at higher throughput.
+enum {
+  kUSBLogBufferSize = 0x800u,
+};
+
+// Total number of logging buffers to be shared among all active streams.
+enum {
+  kUSBLogNumBuffers = 8u,
+};
+
+// Maximum delay before logging data is flushed upstream to the host;
+// microseconds.
+enum {
+  kUSBLogFlushInterval = 500u * 1000u,
+};
+
+// Maximum length of the configuration descriptor.
+enum {
+  kCfgDscrLenMax =
+      (USB_CFG_DSCR_LEN +
+       kUSBLogMaxStreams * (USB_INTERFACE_DSCR_LEN + 1u * USB_EP_DSCR_LEN)),
+};
```

Comment:
Updated. I'm aware that separate enums are not needed; I'd argue that enums are a bad coding style choice because they don't guarantee size or signedness, and that placing unrelated constants into a single enumeration really isn't why the 'enum' language feature exists, but I've changed the code for conformance.

Reply:
> I'd argue that enums are a bad coding style choice because they don't guarantee size or signedness, and that placing unrelated constants into a single enumeration really isn't why the 'enum' language feature exists

Yeah, it's an ~abusive~ creative repurposing of enums but having compile-time constants without involving the preprocessor is nice :)

Regarding the size and signedness, doesn't the syntax `enum : uint32_t { ... }` cover that, or am I misunderstanding you? I suppose we could add to the style guide to always explicitly specify the underlying enumeration type, if that's an issue.

Reply:
Interesting; I hadn't actually encountered that syntax. Thanks! Researching, though, it appears to be C++ only at present, and scheduled for inclusion in the next C standard.

Reply:
In practice it's been supported for a while, even when you select an old `-std=...` version, and our toolchain supports it, so if we want we can start using it.